tools/bss_reordering.py

```diff
@@ -0,0 +1,207 @@
+    build = open(f"build/{version}/oot-{version}.z64", "rb")
+
+    # Find all pointers with different values
+    pointers = []
+    for mapfile_segment in mapfile_segments:
+        for file in mapfile_segment:
+            if not str(file.filepath).endswith(".o"):
+                continue
+            for reloc in read_relocs(file.filepath, file.sectionType):
+                if reloc.offset_32 is not None:
+                    base_value = read_u32(base, file.vrom + reloc.offset_32)
+                    build_value = read_u32(build, file.vrom + reloc.offset_32)
+                elif reloc.offset_hi16 is not None and reloc.offset_lo16 is not None:
+                    if read_u16(base, file.vrom + reloc.offset_hi16) != read_u16(
+                        build, file.vrom + reloc.offset_hi16
+                    ) or read_u16(base, file.vrom + reloc.offset_hi16) != read_u16(
+                        build, file.vrom + reloc.offset_hi16
+                    ):
+                        print(
+                            f"Error: Relocation for {reloc.name} in {file.filepath} references a shifted portion of the ROM.\n"
+                            "Please ensure that the only differences between the baserom and the current build are due to BSS reordering.",
+                            file=sys.stderr,
+                        )
+                        sys.exit(1)
+
+                    base_value = (
+                        read_u16(base, file.vrom + reloc.offset_hi16 + 2) << 16
+                    ) + read_s16(base, file.vrom + reloc.offset_lo16 + 2)
+                    build_value = (
+                        read_u16(build, file.vrom + reloc.offset_hi16 + 2) << 16
+                    ) + read_s16(build, file.vrom + reloc.offset_lo16 + 2)
+                else:
+                    assert False, "Invalid relocation"
+
+                if base_value != build_value:
+                    pointers.append(
+                        Pointer(reloc.name, reloc.addend, base_value, build_value)
+                    )
+
+    # Remove duplicates and sort by address
+    pointers = list({p.base_value: p for p in pointers}.values())
+    pointers.sort(key=lambda p: p.base_value)
+
+    # Go through BSS sections and report differences
+    i = 0
+    for mapfile_segment in mapfile_segments:
+        for file in mapfile_segment:
+            if not file.sectionType == ".bss":
+                continue
+
+            pointers_in_section = [
+                p for p in pointers if file.vram <= p.base_value < file.vram + file.size
+            ]
+            if not pointers_in_section:
+                continue
+
+            print(f"{file.filepath} BSS is reordered:")
+            for i, p in enumerate(pointers_in_section):
+                if i > 0 and p.build_value < pointers_in_section[i - 1].build_value:
+                    print("  --------------------")  # BSS wraps around
+                addend_str = f"+0x{p.addend:X}" if p.addend else ""
```

Comment:
The addend can be signed, which this formatting doesn't handle

Reply:
Nice catch, thanks